### PR TITLE
616 domain knowledge facts rules should be loadable via the rest api

### PR DESCRIPTION
**Pull request by bnouwt**

### CODE REVIEW

File: smart-connector-rest-server/src/main/resources/openapi-sc.yaml [hunk 1]
```diff
@@ -536,6 +536,46 @@ paths:
             application/json; charset=UTF-8:
               schema:
                 $ref: '#/components/schemas/ResponseMessage'
+  /sc/knowledge:
+    post:
+      summary: Use the specified domain knowledge for the given knowledge base during activating proactive (i.e. ASk/POST) knowledge interactions. Using additional domain knowledge can increase semantic interoperability. Note that a sufficient reasoner level determines whether the domain knowledge is actually used.
```
Comment (Sophietje) [inline smart-connector-rest-server/src/main/resources/openapi-sc.yaml:541]: "ASk" -> "ASK"

**Review by Sophietje — CHANGES_REQUESTED** [review]
Two additional comments:
1. I'd recommend adding some tests to test edge cases:
   a) try to load an empty string/file
   b) try and load a file with erroneous syntax
   c) try and load a file with valid syntax but that contains something that we don't support (if that's even a thing, if so we should mention it in the docs)

2. Do we support all syntax from Apache Jena's rule syntax? I saw something about the language tag somewhere. If not, we should include what syntax is supported in the documentation

File: docs/docs/reasoning.md [hunk 2]
```diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
...
+@prefix saref: <https://saref.etsi.org/core/> .
+
+-> ( saref:Sensor rdfs:subClassOf saref:Device ) .
+
+(?x rdfs:subClassOf ?y), (?a rdf:type ?x) -> (?a rdf:type ?y) .
+
+```
+
+:::tip
+Note that the domain fact `saref:Sensor rdfs:subClassOf saref:Device` above is represented as a body-less rule.
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply docs/docs/reasoning.md:199]: Yes, I've added clarification 👍

File: docs/docs/reasoning.md [hunk 3]
```diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
...
+:::
+
+Loading the above domain knowledge into a smart connector will allow its reasoner to derive that every `saref:Sensor` in the network is also a `saref:Device` and whenever a device is requested using the graph pattern `?d rdf:type saref:Device` it will also return data from KIs with the `?s rdf:type saref:Sensor` graph pattern. The default domain knowledge for every smart connector created in a KE Runtime can be configured using the `ke.domain.knowledge.path` configuration property. See [configuration](https://github.com/TNO/knowledge-engine?tab=readme-ov-file#configuration) section for more info.
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply docs/docs/reasoning.md:202]: Yes, level 2. I've added a tip about this with a reference to more info.

File: docs/docs/reasoning.md [hunk 4]
```diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
...
+```java
+smartConnector.setDomainKnowledge(someDomainKnowledge);
+```
+
+The `someDomainKnowledge` variable should be a set of `eu.knowledge.engine.reasoner.Rule` objects. The `eu.knowledge.engine.reasoner.util.JenaRules` class provides some helper methods for creating these objects from files/strings.
+
+</TabItem>
+<TabItem value="bash" label="Rest API">
+
+When using the REST API, you can load the domain knowledge into a specific smart connector using the [`POST /sc/knowledge/`](https://github.com/TNO/knowledge-engine/blob/master/smart-connector-rest-server/src/main/resources/openapi-sc.yaml#L539) operation. The body of the request should contain `text/plain` text following the Apache Jena Rules spec mentioned above:
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply docs/docs/reasoning.md:216]: Yes, I've added a similar phrase before the Java API.

File: reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+import org.apache.jena.reasoner.rulesys.Node_RuleVariable;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.serializer.SerializationContext;
+import org.apache.jena.sparql.util.FmtUtils;
+
+import eu.knowledge.engine.reasoner.BaseRule;
+import eu.knowledge.engine.reasoner.Rule;
+import eu.knowledge.engine.reasoner.api.TriplePattern;
+
+public class JenaRules {
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:25]: Good idea, I've added short description.

File: docs/docs/reasoning.md [hunk 6]
```diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
...
+-> ( saref:Sensor rdfs:subClassOf saref:Device ) .
+
+(?x rdfs:subClassOf ?y), (?a rdf:type ?x) -> (?a rdf:type ?y) .
+```
+
+
+</TabItem>
+</Tabs>
+
+
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply docs/docs/reasoning.md:230]: Agree, I've added it to the docs.

File: reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+	 * Every fact present in the turtle file will be converted to the following
+	 * format:
+	 * 
+	 * <pre>
+	 * {@code
+	 * -> ( <subject> <predicate> <object> ) .
+	 * }
+	 * </pre>
+	 * 
+	 * Note that literals with language tags are not supported.
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:95]: It will not happen often and is actually a characteristic of the Apache Jena Rules language, so if we say we use that specification then the language tag non-support is included. But I think adding it to the documentation is still a good idea. Done!

File: reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+	 * Note that all fact rules (i.e. rules without an antecedent and only a
+	 * consequent) are collected and created into a single data rule suitable for
+	 * the reasoner.
+	 * 
+	 * @param someRules A string of rules in the Apache Jena Rules syntax.
+	 * @return A set of rules that can be used with the KE reasoner.
+	 */
+	public static Set<BaseRule> convertJenaToKeRules(String someRules) {
+		List<org.apache.jena.reasoner.rulesys.Rule> jenaRules = org.apache.jena.reasoner.rulesys.Rule
+				.parseRules(someRules);
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:130]: Good catch 👍 The `ParserException` will actually propagate all the way to the [REST API](https://github.com/TNO/knowledge-engine/pull/624/files#diff-d81c65582ed725549c9323413bc56f7a9dfee1ab1ce26b950886b2b39e8eff9cR276) and is catched there to give a proper error message to the user.

The Java API does not use the Jena Rules syntax, so there we do not have to deal with this `ParserException`.

File: reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+			Set<TriplePattern> consequent = new HashSet<>();
+			for (ClauseEntry ce : r.getHead()) {
+				consequent.add(convertToTriplePattern(((org.apache.jena.reasoner.TriplePattern) ce)));
+			}
+
+			if (antecedent.isEmpty()) {
+				// create KE data rule from consequent
+
+				assert consequent.size() == 1;
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:150]: Good point, I've tested it and if the consequent is 2 the assertion fails and throws an exception that is not catched. With assertions disabled (which is typically the case in production), the `consequent.iterator().next()` will throw a `NoSuchElementException` that also breaks the knowledge engine.

I noticed that the Apache Jena Rules syntax indeed allows multiple consequents although the antecedent is empty, so would:

```
-> (rdf:type      rdfs:range rdfs:Class) (rdfs:Resource  rdf:type  rdfs:Class).
```

and

```
-> (rdf:type      rdfs:range rdfs:Class).
-> (rdfs:Resource  rdf:type  rdfs:Class).
```

have the same semantics? In that case I can just deal with them quite easily. For now I've assumed that and changed the code to deal with multiple heads. Also added a unit test to check this.

**Comment by bnouwt** [comment]
> Two additional comments:
> 
> 1. I'd recommend adding some tests for edge cases:
>    a) try to load an empty string/file
>    b) try and load a file with erroneous syntax
>    c) try and load a file with valid syntax but that contains something that we don't support (if that's even a thing, if so we should mention it in the docs)
> 2. Do we support all syntax from Apache Jena's rule syntax? I saw something about the language tag somewhere. If not, we should include what syntax is supported in the documentation

Thanks for the review @Sophietje! A lot of good comments. Regarding the above comments:

1a) Added this case to `eu.knowledge.engine.reasoner.api.JenaRuleConversionTest`.
1b) This is already tested in `eu.knowledge.engine.rest.api.TestDomainKnowledge.testDomainKnowledge()`.
1c) I've added tests and logic to deal with [Builtin primitives](https://jena.apache.org/documentation/inference/#RULEbuiltins) which the TKE does not support.
2) I've added a tip about language tags in the documentation 👍

File: reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java [hunk 10]
```diff
@@ -165,7 +179,9 @@ public static Set<BaseRule> convertJenaToKeRules(String someRules) {
 			}
 		}
 
-		if (stringRows.size() > 0) {
+		if (stringRows.size() > 0)
+
+		{
```
[the remark this answers is not in the document]
Comment (bnouwt) [reply reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:184]: 👍

**Review by Sophietje — APPROVED** [review]
